pki/pki.go

```diff
@@ -121,12 +129,22 @@ type StatusInfo struct {
 	FailInfo     asn1.BitString `asn1:"optional"`
 }
 
-// ParseFailInfo parses the FailInfo field of a PKIStatusInfo to PKIFailureInfo
-func (psi StatusInfo) ParseFailInfo() (FailureInfo, error) {
-	for _, pfi := range FailureInfos {
-		if psi.FailInfo.At(int(pfi)) != 0 {
-			return pfi, nil
+// Err return nil when si Status is StatusGranted or StatusGrantedWithMods
+//
+// Otherwise, Err returns an error with FailInfo if any.
+func (si StatusInfo) Err() error {
+	var err error
+	if si.Status != StatusGranted && si.Status != StatusGrantedWithMods {
+		for _, fi := range failureInfos {
+			if si.FailInfo.At(int(fi)) != 0 {
+				err = errors.Join(err, fi.Error())
+			}
+		}
+		if err != nil { // there is FailInfo
+			errMsg := strings.ReplaceAll(err.Error(), "\n", "; ")
+			return fmt.Errorf("invalid response with status code %d: %s. Failure info: %s", si.Status, si.Status.String(), errMsg)
```

Comment:
It is better to embed the original errors. If the error message is not good enough, you may define your own `joinError`.

Reply:
In that case, then we don't need to use the `Join()` function here. Will create a new error structure for it.